src/components/tl/mlx5/tl_mlx5_team.c

```diff
@@ -169,40 +146,169 @@ static inline ucc_status_t ucc_tl_mlx5_a2a_team_test(ucc_base_team_t *team)
 
 ucc_status_t ucc_tl_mlx5_team_create_test(ucc_base_team_t *team)
 {
-    ucc_tl_mlx5_team_t *tl_team      = ucc_derived_of(team, ucc_tl_mlx5_team_t);
-    ucc_status_t        a2a_status   = UCC_OK;
-    ucc_status_t        mcast_status = UCC_OK;
+    ucc_tl_mlx5_team_t            *tl_team      = ucc_derived_of(team, ucc_tl_mlx5_team_t);
+    ucc_team_t                    *core_team    = UCC_TL_CORE_TEAM(tl_team);
+    ucc_subset_t                   subset       = {.map = UCC_TL_TEAM_MAP(tl_team),
+                                                .myrank = UCC_TL_TEAM_RANK(tl_team)};
+    ucc_status_t                   a2a_status   = UCC_OK;
+    ucc_status_t                   mcast_status = UCC_OK;
+    ucc_tl_mlx5_mcast_coll_comm_t *comm         = NULL;
+    ucc_status_t                   status;
+
+
+    if (tl_team->global_sync_in_progress) {
+        status = ucc_service_coll_test(tl_team->scoll_req);
+        if (status < 0) {
+            tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                     "failure during service coll exchange: %s",
+                     ucc_status_string(status));
+            return status;
+        }
+        if (UCC_INPROGRESS == status) {
+            return status;
+        }
+
+        ucc_service_coll_finalize(tl_team->scoll_req);
+
+        tl_team->global_sync_in_progress = 0;
+
+        if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_INIT &&
+            tl_team->a2a_state == TL_MLX5_TEAM_STATE_INIT ) {
+            tl_team->a2a_status.global = tl_team->global_status_array[0];
+            tl_team->a2a_state         = TL_MLX5_TEAM_STATE_ALLTOALL_INIT;
```

Comment:
I am a bit confused between the different state enum, what is the high level idea? Like here why do we move from
`tl_team->a2a_state = TL_MLX5_TEAM_STATE_INIT` to `tl_team->a2a_state = TL_MLX5_TEAM_STATE_ALLTOALL_INIT` ?

Reply:
Previously (before this patch) we had two more states where we do allreduce for allltoall. Before doing allreduce, the state was TL_MLX5_TEAM_STATE_INIT and after allreduce is done, the state becomes TL_MLX5_TEAM_STATE_ALLTOALL_INIT.  This is why here we are trying to maintain the same way.

Reply:
So the same was done before?
I don't know, but anyway there seems to be unnecessary complexity here, don't you agree ?

Reply:
it is indeed complex because we are trying to do team create in a nonblocking manner. If we were free to make it blocking, no state machine was needed. in this specific case, I made some changes. Please let me know how it looks in the new patch.

Reply:
looks good, thanks